gnucash/import-export/import-main-matcher.c

```diff
@@ -1435,38 +1496,141 @@ void gnc_gen_trans_list_add_trans_with_ref_id (GNCImportMainMatcher *gui, Transa
+        // It's much faster to gather the imported transactions into a GSList than directly into the
+        // treeview.
+        gui->temp_trans_list = g_slist_prepend (gui->temp_trans_list, transaction_info);
+    }
+    return;
+}
+
+// This creates lists of all splits that could match one of the imported transactions based on their
+// account and date and organized per account
+static GHashTable*
+create_list_of_potential_matches (GNCImportMainMatcher *gui, GtkTreeModel* model, Query *query, gint match_date_hardlimit)
+{
+    GList* query_return = NULL;
+    Account *import_trans_account;
+    time64 import_trans_time, min_time=G_MAXINT64, max_time=0;
+    GList* all_accounts = NULL;
+    static const int secs_per_day = 86400;
+    GSList* imported_trans;
+    GList* potential_match;
+    GHashTable* lists_per_accounts = g_hash_table_new_full(g_direct_hash, g_direct_equal, NULL, (GDestroyNotify)g_slist_free);
+
+    // Go through all imported transactions, gather the list of accounts, and min/max date range.
+    for (imported_trans=gui->temp_trans_list; imported_trans!=NULL; imported_trans=imported_trans->next)
+    {
+        GNCImportTransInfo* transaction_info;
+        transaction_info = imported_trans->data;
+        import_trans_account = xaccSplitGetAccount (gnc_import_TransInfo_get_fsplit (transaction_info));
+        all_accounts = g_list_prepend (all_accounts, import_trans_account);
+        import_trans_time = xaccTransGetDate (gnc_import_TransInfo_get_trans (transaction_info));
+        min_time = MIN (min_time, import_trans_time);
+        max_time = MAX (max_time, import_trans_time);
+    }
 
-        gnc_import_TransInfo_init_matches (transaction_info,
-                                           gui->user_settings);
+    // Make a query to find splits with the right accounts and dates.
+    qof_query_set_book (query, gnc_get_current_book());
+    xaccQueryAddAccountMatch (query, all_accounts,
+                              QOF_GUID_MATCH_ANY, QOF_QUERY_AND);
+    xaccQueryAddDateMatchTT (query,
+                             TRUE, min_time - match_date_hardlimit * secs_per_day,
+                             TRUE, max_time + match_date_hardlimit * secs_per_day,
+                             QOF_QUERY_AND);
+    query_return = qof_query_run (query);
+    g_list_free (all_accounts);
+    
+    // Now put all potential matches into a hash table based on their account.
+    for (potential_match=query_return; potential_match!=NULL; potential_match=potential_match->next)
+    {
+        Account* split_account;
+        GSList* per_account_list;
+        if (gnc_import_split_has_online_id (potential_match->data))
+            continue;
+        split_account = xaccSplitGetAccount (potential_match->data);
+        // This will be NULL the first time around, which is fine for a GSList
+        g_hash_table_steal_extended (lists_per_accounts, import_trans_account, NULL, (gpointer)&per_account_list);
```

Comment:
This fails on ubuntu 18.04 -- it has glib 2.56.4 < 2.58 doesn't have `g_hash_table_steal_extended` -- how annoying!

Reply:
Well, fooey. So it has to be either the original code or a `g_hash_table_lookup();g_hash_table_steal()` two-step. If the original code then use `g_hash_table_foreach_remove` to traverse the hash table only once for destruction.